README.md

![License](https://img.shields.io/npm/l/electron-pos-printer)
![Version](https://img.shields.io/npm/v/electron-pos-printer?label=version)
![Issues](https://img.shields.io/github/issues/Hubertformin/electron-pos-printer)

# Electron-pos-printer
A customizable electron.js printing plugin specifically designed for thermal receipt printers.
It supports 58mm - 80mm. It currently supports versions of electron >= 6.x.x

### Installation
```bash
$ npm install electron-pos-printer
$ yarn add electron-pos-printer
```
### Demo
Check out this [Demo](https://github.com/fssonca/electron-printer ) by [fssonca](https://github.com/fssonca)

### Usage
### In main process
```js
const {PosPrinter} = require("electron-pos-printer");
```
### In render process
```js
const {PosPrinter} = require('electron').remote.require("electron-pos-printer");
```
Electron >= `v10.x.x`
```js
const {PosPrinter} = require('@electron/remote').remote.require("electron-pos-printer");
```

### Example code
```js
const {PosPrinter} = require("electron-pos-printer");
const path = require("path");

const options = {
   preview: false,               // Preview in window or print
   width: '170px',               //  width of content body
   margin: '0 0 0 0',            // margin of content body
   copies: 1,                    // Number of copies to print
   printerName: 'XP-80C',        // printerName: string, check with webContent.getPrinters()
   timeOutPerLine: 400,          // Timeout per line in miliseconds
   pageSize: {                   // The size of the to-be printed page
       width: 220,
       height: 1200 
   } 
}

const data = [
    {
        type: 'image',
        url: 'https://randomuser.me/api/portraits/men/43.jpg',     // file path
        position: 'center',                                  // position of image: 'left' | 'center' | 'right'
        width: '160px',                                           // width of image in px; default: auto
        height: '60px',                                          // width of image in px; default: 50 or '50px'
    },{
        type: 'text',                                       // 'text' | 'barCode' | 'qrCode' | 'image' | 'table
        value: 'SAMPLE HEADING',
        style: {fontWeight: "700", textAlign: 'center', fontSize: "24px"}
    },{
        type: 'text',                       // 'text' | 'barCode' | 'qrCode' | 'image' | 'table'
        value: 'Secondary text',
        style: {textDecoration: "underline", fontSize: "10px", textAlign: "center", color: "red"}
    },{
        type: 'barCode',
        value: '023456789010',
        height: 40,                     // height of barcode, applicable only to bar and QR codes
        width: 2,                       // width of barcode, applicable only to bar and QR codes
        displayValue: true,             // Display value below barcode
        fontsize: 12,
    },{
        type: 'qrCode',
        value: 'https://github.com/Hubertformin/electron-pos-printer',
        height: 55,
        width: 55,
        style: { margin: '10 20px 20 20px' }
    },{
        type: 'table',
        // style the table
        style: {border: '1px solid #ddd'},
        // list of the columns to be rendered in the table header
        tableHeader: ['Animal', 'Age'],
        // multi dimensional array depicting the rows and columns of the table body
        tableBody: [
            ['Cat', 2],
            ['Dog', 4],
            ['Horse', 12],
            ['Pig', 4],
        ],
        // list of columns to be rendered in the table footer
        tableFooter: ['Animal', 'Age'],
        // custom style for the table header
        tableHeaderStyle: { backgroundColor: '#000', color: 'white'},
        // custom style for the table body
        tableBodyStyle: {'border': '0.5px solid #ddd'},
        // custom style for the table footer
        tableFooterStyle: {backgroundColor: '#000', color: 'white'},
    },{
        type: 'table',
        style: {border: '1px solid #ddd'},             // style the table
        // list of the columns to be rendered in the table header
        tableHeader: [{type: 'text', value: 'People'}, {type: 'image', path: path.join(__dirname, 'icons/animal.png')}],
        // multi-dimensional array depicting the rows and columns of the table body
        tableBody: [
            [{type: 'text', value: 'Marcus'}, {type: 'image', url: 'https://randomuser.me/api/portraits/men/43.jpg'}],
            [{type: 'text', value: 'Boris'}, {type: 'image', url: 'https://randomuser.me/api/portraits/men/41.jpg'}],
            [{type: 'text', value: 'Andrew'}, {type: 'image', url: 'https://randomuser.me/api/portraits/men/23.jpg'}],
            [{type: 'text', value: 'Tyresse'}, {type: 'image', url: 'https://randomuser.me/api/portraits/men/53.jpg'}],
        ],
        // list of columns to be rendered in the table footer
        tableFooter: [{type: 'text', value: 'People'}, 'Image'],
        // custom style for the table header
        tableHeaderStyle: { backgroundColor: 'red', color: 'white'},
        // custom style for the table body
        tableBodyStyle: {'border': '0.5px solid #ddd'},
        // custom style for the table footer
        tableFooterStyle: {backgroundColor: '#000', color: 'white'},
    },
]

PosPrinter.print(data, options)
 .then(console.log)
 .catch((error) => {
    console.error(error);
  });

```

## Typescript
### Usage

```typescript
import {PosPrinter, PosPrintData, PosPrintOptions} from "electron-pos-printer";
import * as path from "path";

const options: PosPrintOptions = {
   preview: false,
   width: '170px',       
   margin: '0 0 0 0',    
   copies: 1,
   printerName: 'XP-80C',
   timeOutPerLine: 400,
   pageSize: { height: 301000, width: 71000 } // page size
}

const data: PosPrintData[] = [
    {
        type: 'image',
        url: 'https://randomuser.me/api/portraits/men/43.jpg',     // file path
        position: 'center',                                  // position of image: 'left' | 'center' | 'right'
        width: '160px',                                           // width of image in px; default: auto
        height: '60px',                                          // width of image in px; default: 50 or '50px'
    },{
        type: 'text',                                       // 'text' | 'barCode' | 'qrCode' | 'image' | 'table
        value: 'SAMPLE HEADING',
        style: {fontWeight: "700", textAlign: 'center', fontSize: "24px"}
    },{
        type: 'text',                       // 'text' | 'barCode' | 'qrCode' | 'image' | 'table'
        value: 'Secondary text',
        style: {textDecoration: "underline", fontSize: "10px", textAlign: "center", color: "red"}
    },{
        type: 'barCode',
        value: '023456789010',
        height: 40,                     // height of barcode, applicable only to bar and QR codes
        width: 2,                       // width of barcode, applicable only to bar and QR codes
        displayValue: true,             // Display value below barcode
        fontsize: 12,
    },{
        type: 'qrCode',
        value: 'https://github.com/Hubertformin/electron-pos-printer',
        height: 55,
        width: 55,
        style: { margin: '10 20px 20 20px' }
    },{
        type: 'table',
        // style the table
        style: {border: '1px solid #ddd'},
        // list of the columns to be rendered in the table header
        tableHeader: ['Animal', 'Age'],
        // multi dimensional array depicting the rows and columns of the table body
        tableBody: [
            ['Cat', 2],
            ['Dog', 4],
            ['Horse', 12],
            ['Pig', 4],
        ],
        // list of columns to be rendered in the table footer
        tableFooter: ['Animal', 'Age'],
        // custom style for the table header
        tableHeaderStyle: { backgroundColor: '#000', color: 'white'},
        // custom style for the table body
        tableBodyStyle: {'border': '0.5px solid #ddd'},
        // custom style for the table footer
        tableFooterStyle: {backgroundColor: '#000', color: 'white'},
    },{
        type: 'table',
        style: {border: '1px solid #ddd'},             // style the table
        // list of the columns to be rendered in the table header
        tableHeader: [{type: 'text', value: 'People'}, {type: 'image', path: path.join(__dirname, 'icons/animal.png')}],
        // multi-dimensional array depicting the rows and columns of the table body
        tableBody: [
            [{type: 'text', value: 'Marcus'}, {type: 'image', url: 'https://randomuser.me/api/portraits/men/43.jpg'}],
            [{type: 'text', value: 'Boris'}, {type: 'image', url: 'https://randomuser.me/api/portraits/men/41.jpg'}],
            [{type: 'text', value: 'Andrew'}, {type: 'image', url: 'https://randomuser.me/api/portraits/men/23.jpg'}],
            [{type: 'text', value: 'Tyresse'}, {type: 'image', url: 'https://randomuser.me/api/portraits/men/53.jpg'}],
        ],
        // list of columns to be rendered in the table footer
        tableFooter: [{type: 'text', value: 'People'}, 'Image'],
        // custom style for the table header
        tableHeaderStyle: { backgroundColor: 'red', color: 'white'},
        // custom style for the table body
        tableBodyStyle: {'border': '0.5px solid #ddd'},
        // custom style for the table footer
        tableFooterStyle: {backgroundColor: '#000', color: 'white'},
    },
]

PosPrinter.print(data, options)
 .then(console.log)
 .catch((error) => {
    console.error(error);
  });
```

## Printing options
| Options        | Required | Type                 | Description                                                                                                                                                                                                                               |
|----------------|:--------:|:---------------------|:------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| printerName    |    No    | `string`             | The printer's name. If not set, the system's default printer will be used.                                                                                                                                                                | 
| copies         |    No    | `number`             | The number of copies to print                                                                                                                                                                                                             |
| preview        |    No    | `boolean`            | Preview print job in a window. Default is `false`                                                                                                                                                                                         |
| width          |    No    | `string`             | Width of a page's content                                                                                                                                                                                                                 |
| margin         |    No    | `string`             | Margin of a page's content. CSS margin values can be used Ex. `0 10px`                                                                                                                                                                    | 
| timeOutPerLine |    No    | `number`             | Timeout per line in milliseconds. Default value is 400                                                                                                                                                                                    | 
| silent         |    No    | `boolean`            | To print silently without the system displaying the printer selection window. Default value is `true`                                                                                                                                     | 
| pageSize       |    No    | `SizeOptions`        | Specify the width and height of the print out page                                                                                                                                                                                        |
| pathTemplate   |    No    | `string`             | Path to custom html template. Can be used for custom print styles.                                                                                                                                                                        | 
| header         |    No    | `string`             | Text to be printed as page header.                                                                                                                                                                                                        | 
| footer         |    No    | `string`             | Text to be printed as page footer.                                                                                                                                                                                                        | 
| margins        |    No    | `object`             | Page margins. [See electron docs](https://www.electronjs.org/docs/latest/api/web-contents#contentsprintoptions-callback)                                                                                                                  | 
| landscape      |    No    | `boolean`            | Whether the page should be printed in landscape mode. Default is `false`.                                                                                                                                                                 | 
| scaleFactor    |    No    | `number`             | The scale factor of the web page.                                                                                                                                                                                                         | 
| pagesPerSheet  |    No    | `number`             | The number of pages to print per page sheet.                                                                                                                                                                                              | 
| collate        |    No    | `boolean`            | Whether the page should be collated.                                                                                                                                                                                                      | 
| pageRanges     |    No    | `object[]`           | The page range to print. [See electron docs](https://www.electronjs.org/docs/latest/api/web-contents#contentsprintoptions-callback)                                                                                                       | 
| duplexMode     |    No    | `string`             | Set the duplex mode of the printed web page. [See electron docs](https://www.electronjs.org/docs/latest/api/web-contents#contentsprintoptions-callback)                                                                                   | 
| pageSize       |    No    | `string` or `object` | Specify page size of the printed document. Can be A3, A4, A5, Legal, Letter, Tabloid or an Object containing height and width. [See electron docs](https://www.electronjs.org/docs/latest/api/web-contents#contentsprintoptions-callback) | 
| dpi            |    No    | `object`             | [See electron docs](https://www.electronjs.org/docs/latest/api/web-contents#contentsprintoptions-callback)                                                                                                                                |

## The Print data object
> ### Important
> The `css` property is no longer supported, use the style property instead of css. <br />
> Example: `style: {fontWeight: "700", textAlign: 'center', fontSize: "24px"}`

| Property         | Type                                     | Description                                                                               |
|------------------|------------------------------------------|:------------------------------------------------------------------------------------------|
| type             | `string`                                 | `text`, `qrCode`, `barCode`, `image`, `table` // type `text` can be an html string        |
| value            | `string`                                 | Value of the current row                                                                  |
| height           | `number`                                 | Applicable to type barCode and qrCode                                                     |
| width            | `number`                                 | Applicable to type barCode and qrCode                                                     |
| style            | `PrintDataStyle`                         | Add css styles to line (jsx syntax) <br />ex: `{fontSize: 12, backgroundColor: '#2196f3}` |
| displayValue     | `boolean`                                | Display value of barcode below barcode                                                    |
| position         | `string`                                 | `left`, `center`, `right` applicable to types qrCode and image                            |
| path             | `string`                                 | Path or url to the image asset                                                            |
| url              | `string`                                 | Url to image or a base 64 encoding of image                                               |
| tableHeader      | `PosPrintTableField[]` or `string[]`     | The columns to be rendered in the header of the table. Works with type `table`            |
| tableBody        | `PosPrintTableField[][]` or `string[][]` | The columns to be rendered in the body of the table. Works with type `table`              |
| tableFooter      | `PosPrintTableField[]` or `string[]`     | The columns to rendered it the footer of the table. Works with type `table`               |
| tableHeaderStyle | `string`                                 | Set a custom style to the table header                                                    |
| tableBodyStyle   | `string`                                 | Set a custom style to the table body                                                      |
| tableFooterStyle | `string`                                 | Set a custom style to the table footer                                                    |

## Author
 - Hubert Formin
 - hformin@gmail.com

### Connect with me
<a href="https://www.linkedin.com/in/hformin/" target="_blank">
<img src="https://img.shields.io/badge/LinkedIn-0077B5?style=for-the-badge&logo=linkedin&logoColor=white"  style="margin-bottom: 5px;" />
</a> 
<a href="https://www.instagram.com/ui.hformin/" target="_blank">
<img src="https://img.shields.io/badge/instagram-%ff5851db.svg?color=f02b9a&style=for-the-badge&logo=instagram&logoColor=white" t=instagram style="margin-bottom: 5px;" />
</a>
<a href="https://www.twitter.com/Hformin/" target="_blank">
<img src="https://img.shields.io/badge/Twitter-00ACEE?style=for-the-badge&logo=twitter&logoColor=white"  style="margin-bottom: 5px;" />
</a> 


CHANGELOG.md
# Change Log
## 1.1.0
### Added
- Type image
- Position property for barcode, QR code and image

### Changed
- Fixed extra right margin on page.

## 1.1.1
### Changed
- Fixed long waiting times

## 1.1.2
### Changed
- Fixed images not rendering

## 1.1.3
### Changed
- Fixed early times outs

## 1.2.0
### Added
- Added support for rendering tables

### Changed
- Fixed render process crash


src/lib/body-init.ts
/**
 * Copyright (c) 2022. Author Hubert Formin <hformin@gmail.com>.
 *     This file is called the the render process main html file in lib/pos.html
 *     This page renders data into the view.
 *     Render events are sent from the main process, this process on successful render replies with success of failure event
 */
const fs = require('fs');
const path = require('path');
import {PrintDataStyle} from "../models";
const QRCode = require('qrcode');
const ipcRender = require('electron').ipcRenderer;
const JsBarcode = require("jsbarcode");

type PageElement = HTMLElement | HTMLDivElement | HTMLImageElement;

const body = document.getElementById('main') as HTMLElement;
/**
 * Initialize container in html view, by setting the width and margins specified in the PosPrinter options
 */
ipcRender.on('body-init', function (event, arg) {
    body.style.width = arg?.width || '100%';
    body.style.margin = arg?.margin ||  0;

    event.sender.send('body-init-reply', {status: true, error: null});
});
/**
 * Listen to render event form the main process,
 * Once the main process sends line data, render this data in the web page
 */
ipcRender.on('render-line', (event, arg) => {
    renderDataToHTML(event, arg);
});

/**
 * @function
 * @name generatePageText
 * @param arg {pass argument of type PosPrintData}
 * @description used for type text, used to generate type text
 * */
function generatePageText(arg) {
    const text = arg.value;
    let div = document.createElement('div') as HTMLElement;
    div.innerHTML = text;
    div = applyElementStyles(div, arg.style) as HTMLElement;

    return div;
}
/**
 * @function
 * @name generateTableCell
 * @param arg {pass argument of type PosPrintData}
 * @param type {string}
 * @description used for type text, used to generate type text
 * */
function generateTableCell(arg, type = 'td'): HTMLElement {
    const text = arg.value;

    let cellElement: HTMLElement;

    cellElement = document.createElement(type);
    cellElement.innerHTML  = text
    cellElement = applyElementStyles(cellElement, { padding: '7px 2px', ...arg.style })

    return cellElement;
}
/**
 * @function
 * @name renderImageToPage
 * @param arg {pass argument of type PosPrintData}
 * @description get image from path and return it as a html img
 * */
function renderImageToPage(arg): Promise<HTMLElement> {
    return new Promise(async (resolve, reject) => {
        // Check if string is a valid base64, if yes, send the file url directly
        let uri;
        let img_con = document.createElement('div');

        const image_format = ['apng', 'bmp', 'gif', 'ico', 'cur', 'jpeg', 'jpg', 'jpeg', 'jfif', 'pjpeg',
            'pjp', 'png', 'svg', 'tif', 'tiff', 'webp'];

        img_con = applyElementStyles(img_con, {
            width: '100%',
            display: 'flex',
            justifyContent: arg?.position || 'left'
        }) as HTMLDivElement;

        if (arg.url) {
            if (!isValidHttpUrl(arg.url) && !isBase64(arg.url)) {
                return reject(`Invalid url: ${arg.url}`);
            }

            uri = arg.url;
        } else if (arg.path) {
            // file mut be
            try {
                const data = fs.readFileSync(arg.path);
                let ext = path.extname(arg.path).slice(1);
                if (image_format.indexOf(ext) === -1) {
                    reject(new Error(ext +' file type not supported, consider the types: ' + image_format.join()));
                }
                if (ext === 'svg') { ext = 'svg+xml'; }
                // insert image
                uri = 'data:image/' + ext + ';base64,' + data.toString('base64');
            } catch (e) {
                reject(e);
            }
        }

        let img = document.createElement("img") as HTMLImageElement;

        img = applyElementStyles(img, {
            height: arg.height,
            width: arg.width,
            ...arg.style
        }) as HTMLImageElement;

        img.src = uri;

        // appending
        img_con.prepend(img);
        resolve(img_con);
    });
}

/**
 * @function
 * @name generateTableCell
 * @param str {string}
 * @description Checks if a string is a base64 string
 * */
function isBase64(str) {
    return Buffer.from(str, 'base64').toString('base64') === str;
}

/**
 * @function
 * @name generateTableCell
 * @param element {PageElement}
 * @param style {PrintDataStyle}
 * @description Apply styles to created elements on print web page.
 * */
function applyElementStyles(element: PageElement, style: PrintDataStyle): PageElement {

    for (const styleProp of Object.keys(style)) {
        if (!style[styleProp]) {
            continue;
        }
        element.style[styleProp] = style[styleProp];
    }
    return element;
}

/**
 * @function
 * @name generateTableCell
 * @param url {string}
 * @description Checks is if a string is a valid URL
 * */
function isValidHttpUrl(url: string) {
    let validURL;

    try {
        validURL = new URL(url);
    } catch (_) {
        return false;
    }

    return validURL.protocol === "http:" || validURL.protocol === "https:";
}

/**
 * @function
 * @name generateTableCell
 * @param elementId {string}
 * @param options {object}
 * @description Generate QR in page
 * */
function generateQRCode(elementId: string, { value, height = 15, width = 1}) {
    return new Promise((resolve, reject) => {
        QRCode.toCanvas(
            document.getElementById(elementId), value, {
                width,
                height,
                errorCorrectionLevel: 'H',
                color: '#000',
            }, (error) => {
                if (error) {
                    return reject(error);
                }
                resolve('success!');
            })
    })
}

/**
 * @function
 * @name generatePageText
 * @param event {any} IpcEvent
 * @param arg {pass argument of type PosPrintData}
 * @description Render data as HTML to page
 * */
async function renderDataToHTML(event, arg) {
    switch (arg.line.type) {
        case 'text':
            try {
                body.appendChild(generatePageText(arg.line));
                // sending msg
                event.sender.send('render-line-reply', {status: true, error: null});
            } catch (e) {
                event.sender.send('render-line-reply', {status: false, error: (e as any).toString()});
            }
            return;
        case 'image':
            try {
                const img = await renderImageToPage(arg.line);

                body.appendChild(img);
                event.sender.send('render-line-reply', {status: true, error: null});
            } catch (e) {
                console.log(e)
                event.sender.send('render-line-reply', {status: false, error: (e as any).toString()});
            }
            return;
        case 'qrCode':
            try {
                const container = document.createElement('div');
                container.style.display = 'flex';
                container.style.justifyContent = arg.line?.position || 'left';

                const qrCode = document.createElement('canvas');
                qrCode.setAttribute('id', `qrCode${arg.lineIndex}`);
                applyElementStyles(qrCode, { 'textAlign': arg.line.position ? '-webkit-' + arg.line.position : '-webkit-left'});

                container.appendChild(qrCode);
                body.appendChild(container);

                await generateQRCode(`qrCode${arg.lineIndex}`, {
                    value: arg.line.value,
                    width: arg.line.width,
                    height: arg.line.height,
                });

                // $(`#qrcode${barcodeNumber}`).attr('style',arg.style);
                event.sender.send('render-line-reply', {status: true, error: null});
            } catch(e) {
                event.sender.send('render-line-reply', {status: false, error: (e as any).toString()});
            }
            return;
        case 'barCode':
            try {
                const barcodeEl = document.createElementNS("http://www.w3.org/2000/svg", "svg");
                barcodeEl.setAttributeNS(null, 'id', `barCode-${arg.lineIndex}`);
                body.appendChild(barcodeEl);
                
                JsBarcode(`#barCode-${arg.lineIndex}`, arg.line.value, {
                    // format: "",
                    lineColor: "#000",
                    textMargin: 0,
                    fontOptions: 'bold',
                    fontSize: arg.line.fontsize || 12,
                    width: parseInt(arg.line.width) || 4,
                    height: parseInt(arg.line.height) || 40,
                    displayValue: !!arg.line.displayValue
                });
                // send
                event.sender.send('render-line-reply', {status: true, error: null});
            } catch(e) {
                event.sender.send('render-line-reply', {status: false, error: (e as any).toString()});
            }
            return;
        case 'table':
            // Creating table
            let tableContainer = document.createElement('div');
            tableContainer.setAttribute('id', `table-container-${arg.lineIndex}`)
            let table = document.createElement('table');
            table.setAttribute('id', `table${arg.lineIndex}`);
            table = applyElementStyles(table, { ...arg.line.style }) as HTMLTableElement;
            
            let tHeader = document.createElement('thead');
            tHeader = applyElementStyles(tHeader, arg.line.tableHeaderStyle) as HTMLTableSectionElement;
            
            let tBody = document.createElement('tbody');
            tBody = applyElementStyles(tBody, arg.line.tableBodyStyle) as HTMLTableSectionElement;
            
            let tFooter = document.createElement('tfoot');
            tFooter = applyElementStyles(tFooter, arg.line.tableFooterStyle) as HTMLTableSectionElement;
            // 1. Headers
            if (arg.line.tableHeader) {
                for (const headerArg of arg.line.tableHeader) {
                    {
                        if (typeof headerArg === "object") {
                            switch (headerArg.type) {
                                case 'image':
                                    await renderImageToPage(headerArg)
                                        .then(img => {
                                            const th = document.createElement(`th`);
                                            th.appendChild(img);
                                            tHeader.appendChild(th);
                                        }).catch((e) => {
                                            event.sender.send('render-line-reply', {status: false, error: (e as any).toString()});
                                        })
                                    break;
                                case 'text':
                                    tHeader.appendChild(generateTableCell(headerArg, 'th'));
                                    break;
                            }
                        } else {
                            const th = document.createElement(`th`);
                            th.innerHTML = headerArg;
                            tHeader.appendChild(th);
                        }
                    }
                }
            }
            // 2. Body
            if (arg.line.tableBody) {
                for (const bodyRow of arg.line.tableBody) {
                    const rowTr = document.createElement('tr');
                    for (const colArg of bodyRow) {
                        if (typeof colArg === 'object') {
                            switch (colArg.type) {
                                case 'image':
                                    await renderImageToPage(colArg)
                                        .then(img => {
                                            const th = document.createElement(`td`);
                                            th.appendChild(img);
                                            rowTr.appendChild(th);
                                        }).catch((e) => {
                                            event.sender.send('render-line-reply', {status: false, error: (e as any).toString()});
                                        })
                                    break;
                                case 'text':
                                    rowTr.appendChild(generateTableCell(colArg));
                                    break;
                            }
                        } else {
                            const td = document.createElement(`td`);
                            td.innerHTML = colArg;
                            rowTr.appendChild(td);
                        }
                    }
                    tBody.appendChild(rowTr);
                }
            }
            // 3. Footer
            if (arg.line.tableFooter) {
                for (const footerArg of arg.line.tableFooter)  {
                    if (typeof footerArg === 'object') {
                        switch (footerArg.type) {
                            case 'image':
                                await renderImageToPage(footerArg)
                                    .then(img => {
                                        const footerTh = document.createElement(`th`);
                                        footerTh.appendChild(img);
                                        tFooter.appendChild(footerTh);
                                    }).catch((e) => {
                                        event.sender.send('render-line-reply', {status: false, error: (e as any).toString()});
                                    })
                                break;
                            case 'text':
                                tFooter.appendChild(generateTableCell(footerArg, 'th'));
                                break;
                        }
                    } else {
                        const footerTh = document.createElement(`th`);
                        footerTh.innerHTML = footerArg;
                        tFooter.appendChild(footerTh);
                    }
                }
            }
            // render table
            table.appendChild(tHeader);
            table.appendChild(tBody);
            table.appendChild(tFooter);
            tableContainer.appendChild(table);
            body.appendChild(tableContainer);
            // send
            event.sender.send('render-line-reply', {status: true, error: null});
            return;
    }
}

